### PR TITLE
Added attention matrix support for experiment

### PR DESCRIPTION
**Pull request by TahaZahid05**
The PR aims to add support of manual attention calculation for when we need attention matrix for visualization as well. Implementation reference is taken from: [torch.nn.functional.scaled_dot_product_attention](https://docs.pytorch.org/docs/2.12/generated/torch.nn.functional.scaled_dot_product_attention.html)

It also makes the existing spda function calling code clean by reducing the spda call in the conditional.

### CODE REVIEW

**Comment by aaishwarymishra** [comment]
I think it would be better if we make it a helper method and 1:1 replica of the provided code :)

File: examples/string-reverse-fixed/models/transformer.py [hunk 1]
```diff
@@ -111,29 +112,85 @@ def forward(
...
+            context, attn_weights = scaled_dot_product_attention_reference(
                 q,
                 k,
                 v,
                 attn_mask=mask,
-                is_causal=False,
-                dropout_p=self.dropout_p if self.training else 0.0,
+                dropout_p=self.dropout_p,
+                is_causal=is_causal,
+                training=self.training,
```
Comment (aaishwarymishra) [inline examples/string-reverse-fixed/models/transformer.py:131]: I dont we need extra training, parameter.

Comment (TahaZahid05) [reply examples/string-reverse-fixed/models/transformer.py:131]: ig it's fine since we want to keep it the same as the reference pytorch implementation. The reference function just changes dropout based on training param in function rather than in calling
